cpp/src/arrow/io/file.cc

```diff
@@ -152,20 +152,39 @@ static inline int64_t lseek64_compat(int fd, int64_t pos, int whence) {
 #endif
 }
 
+#if defined(_MSC_VER)
+static inline Status Utf8ToUtf16(const std::string& input, std::wstring &result) {
+    if (input.empty()) {
+        result = std::wstring();
+        return Status::OK();
+    }
+
+    size_t charsNeeded = ::MultiByteToWideChar(CP_UTF8, 0,
+        input.data(), (int)input.size(), NULL, 0);
+    if (charsNeeded == 0) {
+        return Status::IOError("Failed converting UTF-8 string to UTF-16");
+    }
+
+    std::vector<wchar_t> buffer(charsNeeded);
+    size_t charsConverted = ::MultiByteToWideChar(CP_UTF8, 0,
+        input.data(), static_cast<int>(input.size()), &buffer[0], static_cast<int>(buffer.size()));
+    if (charsConverted == 0) {
+        return Status::IOError("Failed converting UTF-8 string to UTF-16");
+    }
+
+    result = std::wstring(&buffer[0], charsConverted);
```

Comment:
It looks like this code is from StackOverflow: http://stackoverflow.com/a/26914562/776560 -- you cannot use StackOverflow code in ASF projects. 

Since we have C++11, I think we should be able to use the locale/codecvt libraries in the STL